src/libNode/MicroBlockProcessing.cpp

```diff
@@ -494,6 +505,171 @@ bool Node::OnCommitFailure([
+            }
+        }
+    };
+
+    auto findOneFromCreated = [this](Transaction& t) -> bool {
+        lock_guard<mutex> g(m_mutexCreatedTransactions);
+
+        auto& listIdx = m_createdTransactions.get<MULTI_INDEX_KEY::GAS_PRICE>();
+        if (!listIdx.size())
+        {
+            return false;
+        }
+
+        auto it = listIdx.begin();
+        t = std::move(*it);
+        listIdx.erase(it);
+        return true;
+    };
+
+    auto appendOne = [this](const Transaction& t) {
+        lock_guard<mutex> g(m_mutexProcessedTransactions);
+        auto& processedTransactions
+            = m_processedTransactions[m_mediator.m_currentEpochNum];
+        processedTransactions.insert(make_pair(t.GetTranID(), t));
+    };
+
+    uint256_t gasUsedTotal = 0;
+
+    while (txn_sent_count < MAXSUBMITTXNPERNODE * m_myShardMembers->size()
+           && gasUsedTotal < MICROBLOCK_GAS_LIMIT)
+    {
+        Transaction t;
+
+        // check m_addrNonceTxnMap contains any txn meets right nonce,
+        // if contains, process it withou increment the txn_sent_count as it's already
+        // incremented when inserting
+        if (findOneFromAddrNonceTxnMap(t))
+        {
+            // check whether m_createdTransaction have transaction with same Addr and nonce
+            // if has and with larger gasPrice then replace with that one. (*optional step)
+            findSameNonceButHigherGasPrice(t);
+
+            uint256_t gasUsed = 0;
+            if (m_mediator.m_validator->CheckCreatedTransaction(t, gasUsed)
+                || !t.GetCode().empty() || !t.GetData().empty())
+            {
+                appendOne(t);
+                gasUsedTotal += gasUsed;
+                continue;
+            }
+        }
+        // if no txn in u_map meet right nonce process new come-in transactions
+        else if (findOneFromCreated(t))
+        {
+            uint256_t gasUsed = 0;
+
+            // check nonce, if nonce larger than expected, put it into m_addrNonceTxnMap
+            if (t.GetNonce()
+                > AccountStore::GetInstance().GetNonceTemp(t.GetSenderAddr())
+                    + 1)
+            {
+                auto it1 = m_addrNonceTxnMap.find(t.GetSenderAddr());
+                if (it1 != m_addrNonceTxnMap.end())
+                {
+                    auto it2 = it1->second.find(t.GetNonce());
+                    if (it2 != it1->second.end())
+                    {
+                        // found the txn with same addr and same nonce
+                        // then compare the gasprice and remains the higher one
+                        if (t.GetGasPrice() > it2->second.GetGasPrice())
+                        {
+                            it2->second = t;
+                        }
+                        txn_sent_count++;
+                        continue;
+                    }
+                }
+                m_addrNonceTxnMap[t.GetSenderAddr()].insert({t.GetNonce(), t});
```

Comment:
Looks like we don't empty m_addrNonceTxnMap at any point yet.  Should we do it at some point, to avoid having the map grow uncontrolled?

Reply:
Done